lib/core/base/audit.js

```diff
@@ -142,48 +160,63 @@ Audit.prototype.addCheck = function (spec) {
  */
 Audit.prototype.run = function (context, options, resolve, reject) {
 	'use strict';
+
 	this.normalizeOptions(options);
 
 	axe._selectCache = [];
-	var q = axe.utils.queue();
-	this.rules.forEach(function (rule) {
-		if (axe.utils.ruleShouldRun(rule, context, options)) {
-			if (options.performanceTimer) {
-				var markEnd = 'mark_rule_end_' + rule.id;
-				var markStart = 'mark_rule_start_' + rule.id;
-				axe.utils.performanceTimer.mark(markStart);
-			}
-			q.defer(function (res, rej) {
-				rule.run(context, options, function(out) {
-					if (options.performanceTimer) {
-						axe.utils.performanceTimer.mark(markEnd);
-						axe.utils.performanceTimer.measure('rule_'+rule.id, markStart, markEnd);
-					}
-					res(out);
-				}, function (err) {
-					if (!options.debug) {
-						var errResult = Object.assign(new RuleResult(rule), {
-							result: axe.constants.CANTTELL,
-							description: 'An error occured while running this rule',
-							message: err.message,
-							stack: err.stack,
-							error: err
-						});
-						res(errResult);
 
-					} else {
-						rej(err);
+	// preload assets if necessary
+	axe.utils.preload(options)
+		.then(([preloadedAssets]) => { // q - retuns array -> destructure/ pluck the first item, is an key value map of preloadedAssets
+			// mutate options with preloadedAssets
+			options = {
+				...options,
+				preloadedAssets
+			};
+			// construct a queue of rules to execute
+			this.rules
+				.reduce((out, rule) => {
+					if (axe.utils.ruleShouldRun(rule, context, options)) {
+						const marker = performanceMarker(rule.id, options.performanceTimer);
+						out.defer((resolve, reject) => {
+							rule.run(
+								context,
+								options,
+								(result) => {
+									marker();
```

Comment:
This is really just a style nit, but this code is very difficult to read due to the "pyramid of doom" here. It might make this easier on the eyes/brain if we pulled some of this logic out of here and into smaller functions.

For example,

```js
// construct a queue of rules to execute
this.rules.reduce(getDeferredRule, axe.utils.queue())
```

Reply:
Nice article describing why it's "easier" to read/write/maintain when code isn't indented this way: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88.